File: scripts/generate_assets.js
import fs from 'fs';
import path from 'path';
import zlib from 'zlib';

/**
 * Creates a raw valid PNG Buffer for a given width, height, and solid RGB color.
 */
export function createPng(width, height, colorRgb = [102, 45, 145]) {
  const [r, g, b] = colorRgb;
  const signature = Buffer.from([0x89, 0x50, 0x4e, 0x47, 0x0d, 0x0a, 0x1a, 0x0a]);

  // IHDR chunk
  const ihdrData = Buffer.alloc(13);
  ihdrData.writeUInt32BE(width, 0);
  ihdrData.writeUInt32BE(height, 4);
  ihdrData.writeUInt8(8, 8); // bit depth 8
  ihdrData.writeUInt8(2, 9); // color type 2 (RGB)
  ihdrData.writeUInt8(0, 10); // compression
  ihdrData.writeUInt8(0, 11); // filter
  ihdrData.writeUInt8(0, 12); // interlace

  const ihdrChunk = createPngChunk('IHDR', ihdrData);

  // IDAT chunk
  const rowLength = 1 + width * 3;
  const rawData = Buffer.alloc(rowLength * height);
  for (let y = 0; y < height; y++) {
    const offset = y * rowLength;
    rawData[offset] = 0; // Filter type 0 (None)
    for (let x = 0; x < width; x++) {
      const pixelOffset = offset + 1 + x * 3;
      rawData[pixelOffset] = r;
      rawData[pixelOffset + 1] = g;
      rawData[pixelOffset + 2] = b;
    }
  }

  const compressedData = zlib.deflateSync(rawData);
  const idatChunk = createPngChunk('IDAT', compressedData);

  // IEND chunk
  const iendChunk = createPngChunk('IEND', Buffer.alloc(0));

  return Buffer.concat([signature, ihdrChunk, idatChunk, iendChunk]);
}

function createPngChunk(type, data) {
  const typeBuf = Buffer.from(type, 'ascii');
  const lenBuf = Buffer.alloc(4);
  lenBuf.writeUInt32BE(data.length, 0);

  const crcBuf = Buffer.alloc(4);
  const crcInput = Buffer.concat([typeBuf, data]);
  const crcValue = zlib.crc32(crcInput);
  crcBuf.writeUInt32BE(crcValue >>> 0, 0);

  return Buffer.concat([lenBuf, typeBuf, data, crcBuf]);
}

export function generateRokuAssets() {
  const assetsDir = path.join(process.cwd(), 'assets', 'images');
  if (!fs.existsSync(assetsDir)) {
    fs.mkdirSync(assetsDir, { recursive: true });
  }

  const assets = [
    { name: 'icon_hd.png', width: 290, height: 218 },
    { name: 'icon_fhd.png', width: 540, height: 360 },
    { name: 'splash_hd.png', width: 1280, height: 720 },
    { name: 'splash_fhd.png', width: 1920, height: 1080 },
  ];

  for (const asset of assets) {
    const filePath = path.join(assetsDir, asset.name);
    if (!fs.existsSync(filePath)) {
      const pngBuf = createPng(asset.width, asset.height, [102, 45, 145]);
      fs.writeFileSync(filePath, pngBuf);
      console.log(`🎨 Generated Roku asset: assets/images/${asset.name} (${asset.width}x${asset.height})`);
    }
  }
}

if (process.argv[1] === new URL(import.meta.url).pathname) {
  generateRokuAssets();
}

File: scripts/generate_assets.py
#!/usr/bin/env python3
import os
import zlib
import struct

def create_png(width, height, color_rgb=(102, 45, 145)):
    """Creates a raw valid PNG byte stream for a given width, height, and solid RGB color."""
    r, g, b = color_rgb
    # PNG signature
    png_sig = b'\x89PNG\r\n\x1a\n'
    
    # IHDR chunk: width (4 bytes), height (4 bytes), bit depth 8, color type 2 (RGB), compression 0, filter 0, interlace 0
    ihdr_data = struct.pack('>IIBBBBB', width, height, 8, 2, 0, 0, 0)
    ihdr_crc = zlib.crc32(b'IHDR' + ihdr_data)
    ihdr_chunk = struct.pack('>I', len(ihdr_data)) + b'IHDR' + ihdr_data + struct.pack('>I', ihdr_crc)
    
    # IDAT chunk (image data)
    # Each row starts with filter byte 0x00, followed by RGB bytes for each pixel
    row_bytes = b'\x00' + bytes([r, g, b] * width)
    raw_data = row_bytes * height
    compressed_data = zlib.compress(raw_data)
    idat_crc = zlib.crc32(b'IDAT' + compressed_data)
    idat_chunk = struct.pack('>I', len(compressed_data)) + b'IDAT' + compressed_data + struct.pack('>I', idat_crc)
    
    # IEND chunk
    iend_crc = zlib.crc32(b'IEND')
    iend_chunk = struct.pack('>I', 0) + b'IEND' + struct.pack('>I', iend_crc)
    
    return png_sig + ihdr_chunk + idat_chunk + iend_chunk

def generate_roku_assets():
    os.makedirs("assets/images", exist_ok=True)
    
    # Roku standard required icon & splash resolutions:
    # HD Icon: 290x218
    # FHD Icon: 540x360
    # Splash FHD: 1920x1080
    assets = [
        ("assets/images/icon_hd.png", 290, 218),
        ("assets/images/icon_fhd.png", 540, 360),
        ("assets/images/splash_fhd.png", 1920, 1080)
    ]
    
    for filepath, w, h in assets:
        if not os.path.exists(filepath):
            png_bytes = create_png(w, h, color_rgb=(102, 45, 145))
            with open(filepath, "wb") as f:
                f.write(png_bytes)
            print(f"🎨 Generated Roku asset: {filepath} ({w}x{h})")

if __name__ == "__main__":
    generate_roku_assets()
